cpp/src/inflight_requests.cpp

```diff
@@ -47,7 +47,19 @@ void InflightRequests::remove(const Request* const request)
       return;
     } else if (result == -1) {
       auto search = _inflightRequests->find(request);
-      if (search != _inflightRequests->end()) _inflightRequests->erase(search);
+      decltype(search->second) tmpRequest;
+      if (search != _inflightRequests->end()) {
+        /**
+         * If this is the last request to hold `std::shared_ptr<ucxx::Endpoint>` erasing it
+         * may cause the `ucxx::Endpoint`s destructor and subsequently the `close()` method
+         * to be called which will in turn call `cancelAll()` and attempt to take the
+         * mutexes. For this reason we should make a temporary copy of the request being
+         * erased from `_inflightRequests` to allow unlocking the mutexes and only then
+         * destroy the object upon this method's return.
+         */
+        tmpRequest = search->second;
+        _inflightRequests->erase(search);
+      }
```

Comment:
No, in `cancelAll` we [don't touch the request(s) anymore after we clear `_inflightRequests`](https://github.com/rapidsai/ucxx/blob/8e8b76470cad44537ee88d31fa5b2e514c612155/cpp/src/inflight_requests.cpp#L73-L75). The issue here is removing the request from `_inflightRequests` and then attempting to access it.

Reply:
Perhaps I am missing something? when we call `clear` that drops all the request references, which might mean that the last request to be dropped is the only holder of a reference to the `Endpoint`. So now, that calls `Endpoint.close()` which in turn would call `cancelAll()`? And we're already holding the locks so that would fail?

Reply:
No, as per usual you're right and I'm wrong. I don't know why but I drifted very far talking of a [different situation](https://github.com/rapidsai/ucxx/blob/8e8b76470cad44537ee88d31fa5b2e514c612155/cpp/src/request.cpp#L112-L116).

The case you're describing doesn't happen currently because `cancelAll()` is only called from `cancelInflightRequests()` which in turn is only called from `Endpoint.close()`. In any case I think the appropriate is to release the lock as soon as possible and let cancelation continue, so if in the future we ever call `cancelAll()` elsewhere and `cancel()` eventually triggers `Endpoint.close()`, we'll not end up in a deadlock and `cancelAll()` will  essentially be a no-op. Done in https://github.com/rapidsai/ucxx/pull/99/commits/397987e10b6cbed97d96a6bb7124c09919f2ab78 .